lgc/patch/Continufy.cpp

```diff
@@ -0,0 +1,196 @@
+static unsigned getReturnedLevels(int32_t stage) {
+  // Traversal will return to RGS.
+  if (stage == -1)
+    return 1;
+  rt::RayTracingShaderStage rtStage = static_cast<rt::RayTracingShaderStage>(stage);
+  switch (rtStage) {
+  case rt::RayTracingShaderStage::ShaderStageRayGeneration:
+    llvm_unreachable("Raygen shader should arrive here.\n");
+  case rt::RayTracingShaderStage::ShaderStageClosestHit:
+  case rt::RayTracingShaderStage::ShaderStageMiss:
+    // Traversal | RGS
+    return (1 << 2 | 1);
+  case rt::RayTracingShaderStage::ShaderStageCallable:
+    // CHS/Miss/Callable | Traversal | RGS
+    return (1 << 2 | 1 << 1 | 1);
+  case rt::RayTracingShaderStage::ShaderStageAnyHit:
+    // IS | Traversal | RGS
+    return (1 << 4 | 1 << 2 | 1);
+  default:
+    llvm_unreachable("Unknown raytracing shader type.\n");
+  }
+}
+
+/// Return level of the ray-tracing stage.
+static unsigned getLevelOfRtStage(int32_t stage) {
+  // Traversal
+  if (stage == -1)
+    return 2;
+  rt::RayTracingShaderStage rtStage = static_cast<rt::RayTracingShaderStage>(stage);
+  switch (rtStage) {
+  case rt::RayTracingShaderStage::ShaderStageRayGeneration:
+    return 0;
+  case rt::RayTracingShaderStage::ShaderStageClosestHit:
+  case rt::RayTracingShaderStage::ShaderStageMiss:
+  case rt::RayTracingShaderStage::ShaderStageCallable:
+    return 1;
+  case rt::RayTracingShaderStage::ShaderStageAnyHit:
+    return 3;
+  case rt::RayTracingShaderStage::ShaderStageIntersection:
+    return 4;
+  default:
+    llvm_unreachable("Unknown raytracing shader type.\n");
+  }
+}
+
+// =====================================================================================================================
+// Executes this LLVM patching pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+// @param [in/out] analysisManager : Analysis manager to use for this transformation
+// @returns : The preserved analyses (The analyses that are still valid after this pass)
+PreservedAnalyses Continufy::run(Module &module, ModuleAnalysisManager &analysisManager) {
+  LLVM_DEBUG(dbgs() << "Run the Continufy pass \n");
+  m_pipelineState = analysisManager.getResult<PipelineStateWrapper>(module).getPipelineState();
+  LLVMContext &context = module.getContext();
+
+  llvm_dialects::Builder builder(context);
+  SmallVector<Instruction *> tobeErased;
+  SmallVector<Function *> functions;
+
+  for (auto &fn : make_early_inc_range(module.functions())) {
+    MDNode *continufyStage = fn.getMetadata("continufy.stage");
+    Function *fnPtr = &fn;
+    int32_t currentRtStage;
+    if (continufyStage) {
+      fnPtr = insertCpsArguments(fn);
+      currentRtStage = mdconst::extract<ConstantInt>(continufyStage->getOperand(0))->getSExtValue();
+      unsigned level = getLevelOfRtStage(currentRtStage);
+      fnPtr->setMetadata("lgc.cps", MDNode::get(context, ConstantAsMetadata::get(builder.getInt32(level))));
```

Comment:
Isn’t there a cps helper for that?

Reply:
I did not realize it. Thanks for letting me know.